### ISSUE TITLE
Gnome-shell crashes when playing youtube video

### ISSUE THREAD

**Issue by fantix** [input-leap/input-leap#1679]
### What happened?

Gnome-shell crashes:
* occassionally when the InputLeap client captures cursor, or
* reliably when playing a youtube video.
* Doesn't happen if InputLeap is stopped.
* Does happen even if all Gnome extensions are turned off.

### Version

From Git HEAD or commit (specify below)

### Git commit hash (if applicable)

fa1bfd99

### If applicable, where did you install Input Leap from?

https://aur.archlinux.org/packages/input-leap-git

### What OSes are you seeing the problem on? (Check all that apply)

Linux

### What OS versions are you using?

Server (68aac941) on macOS while client (fa1bfd99) on ArchLinux (x86_64).

### Relevant log output

```shell
Aug 06 22:56:57 host-pc gnome-shell[4023]: Received an X Window System error.
                                              This probably reflects a bug in the program.
                                              The error was 'BadMatch (invalid parameter attributes)'.
                                                (Details: serial 6011 error_code 8 request_code 147 (unknown) minor_code 6)
                                                (Note to programmers: normally, X errors are reported asynchronously;
                                                 that is, you will receive the error a while after causing it.
                                                 To debug your program, run it with the MUTTER_SYNC environment
                                                 variable to change this behavior. You can then get a meaningful
                                                 backtrace from your debugger if you break on the meta_x_error() function.)
Aug 06 22:56:57 host-pc gnome-shell[4023]: == Stack trace for context 0x5593cdd78cf0 ==
Aug 06 22:56:57 host-pc systemd[1]: Started Process Core Dump (PID 5526/UID 0).
Aug 06 22:56:58 host-pc systemd-coredump[5527]: [🡕] Process 4023 (gnome-shell) of user 1000 dumped core.

                                                   Stack trace of thread 4023:
                                                   #0  0x00007f396e28e83c n/a (libc.so.6 + 0x8e83c)
                                                   #1  0x00007f396e23e668 raise (libc.so.6 + 0x3e668)
                                                   #2  0x00005593ccfb7f19 n/a (gnome-shell + 0x3f19)
                                                   #3  0x00007f396e23e710 n/a (libc.so.6 + 0x3e710)
                                                   #4  0x00007f396ecde018 g_log_structured_array (libglib-2.0.so.0 + 0x61018)
                                                   #5  0x00007f396ecde313 g_log_default_handler (libglib-2.0.so.0 + 0x61313)
                                                   #6  0x00007f396ecdf1c5 g_logv (libglib-2.0.so.0 + 0x621c5)
                                                   #7  0x00007f396ecdf494 g_log (libglib-2.0.so.0 + 0x62494)
                                                   #8  0x00007f396e717e47 n/a (libmutter-12.so.0 + 0x117e47)
                                                   #9  0x00007f396e04568c _XError (libX11.so.6 + 0x4468c)
                                                   #10 0x00007f396e045798 n/a (libX11.so.6 + 0x44798)
                                                   #11 0x00007f396e045855 n/a (libX11.so.6 + 0x44855)
                                                   #12 0x00007f396e0458ea _XEventsQueued (libX11.so.6 + 0x448ea)
                                                   #13 0x00007f396e038238 XPending (libX11.so.6 + 0x37238)
                                                   #14 0x00007f396ecd7f8e g_main_context_check (libglib-2.0.so.0 + 0x5af8e)
                                                   #15 0x00007f396ed34b28 n/a (libglib-2.0.so.0 + 0xb7b28)
                                                   #16 0x00007f396ecd6fef g_main_loop_run (libglib-2.0.so.0 + 0x59fef)
                                                   #17 0x00007f396e6cddcb meta_context_run_main_loop (libmutter-12.so.0 + 0xcddcb)
                                                   #18 0x00005593ccfb7526 n/a (gnome-shell + 0x3526)
                                                   #19 0x00007f396e227cd0 n/a (libc.so.6 + 0x27cd0)
                                                   #20 0x00007f396e227d8a __libc_start_main (libc.so.6 + 0x27d8a)
                                                   #21 0x00005593ccfb7835 n/a (gnome-shell + 0x3835)

                                                   Stack trace of thread 4037:
                                                   #0  0x00007f396e2894ae n/a (libc.so.6 + 0x894ae)
                                                   #1  0x00007f396e28bd40 pthread_cond_wait (libc.so.6 + 0x8bd40)
                                                   #2  0x00007f394dec781c n/a (radeonsi_dri.so + 0xc781c)
                                                   #3  0x00007f394df1472c n/a (radeonsi_dri.so + 0x11472c)
                                                   #4  0x00007f396e28c9eb n/a (libc.so.6 + 0x8c9eb)
                                                   #5  0x00007f396e31123c n/a (libc.so.6 + 0x11123c)

                                                   Stack trace of thread 4038:
                                                   #0  0x00007f396e2894ae n/a (libc.so.6 + 0x894ae)
                                                   #1  0x00007f396e28bd40 pthread_cond_wait (libc.so.6 + 0x8bd40)
                                                   #2  0x00007f394dec781c n/a (radeonsi_dri.so + 0xc781c)
                                                   #3  0x00007f394df1472c n/a (radeonsi_dri.so + 0x11472c)
                                                   #4  0x00007f396e28c9eb n/a (libc.so.6 + 0x8c9eb)
                                                   #5  0x00007f396e31123c n/a (libc.so.6 + 0x11123c)

                                                   Stack trace of thread 4029:
                                                   #0  0x00007f396e3039df __poll (libc.so.6 + 0x1039df)
                                                   #1  0x00007f396ed34c2f n/a (libglib-2.0.so.0 + 0xb7c2f)
                                                   #2  0x00007f396ecd6fef g_main_loop_run (libglib-2.0.so.0 + 0x59fef)
                                                   #3  0x00007f396ef3a28c n/a (libgio-2.0.so.0 + 0x11028c)
                                                   #4  0x00007f396ed09d75 n/a (libglib-2.0.so.0 + 0x8cd75)
                                                   #5  0x00007f396e28c9eb n/a (libc.so.6 + 0x8c9eb)
                                                   #6  0x00007f396e31123c n/a (libc.so.6 + 0x11123c)

                                                   Stack trace of thread 4042:
                                                   #0  0x00007f396e2894ae n/a (libc.so.6 + 0x894ae)
                                                   #1  0x00007f396e28bd40 pthread_cond_wait (libc.so.6 + 0x8bd40)
                                                   #2  0x00007f396d19b591 _ZN7mozilla6detail21ConditionVariableImpl8wait_forERNS0_9MutexImplERKNS_16BaseTimeDurationINS_27TimeDurationValueCalculatorEEE (libmozjs-102.so + 0x159b591)
                                                   #3  0x00007f396d19b492 n/a (libmozjs-102.so + 0x159b492)
                                                   #4  0x00007f396d34d63e n/a (libmozjs-102.so + 0x174d63e)
                                                   #5  0x00007f396e28c9eb n/a (libc.so.6 + 0x8c9eb)
                                                   #6  0x00007f396e31123c n/a (libc.so.6 + 0x11123c)

                                                   Stack trace of thread 4031:
                                                   #0  0x00007f396e2894ae n/a (libc.so.6 + 0x894ae)
                                                   #1  0x00007f396e28bd40 pthread_cond_wait (libc.so.6 + 0x8bd40)
                                                   #2  0x00007f394dec781c n/a (radeonsi_dri.so + 0xc781c)
                                                   #3  0x00007f394df1472c n/a (radeonsi_dri.so + 0x11472c)
                                                   #4  0x00007f396e28c9eb n/a (libc.so.6 + 0x8c9eb)
                                                   #5  0x00007f396e31123c n/a (libc.so.6 + 0x11123c)
```

### Any other information

* xorg-server 21.1.8-2
* gnome-shell 1:44.3-1
* gtk4 1:4.10.5-1
* gtk3 1:3.24.38-1

**Comment by shymega** [comment]
This is a very interesting bug. We need some more context _before_ the X11 error, as indicated.

And to check, you're not building with the `libei` flag, right?

**Comment by fantix** [comment]
Right, I can't see any `libei` flag in the PKGBUILD file. I'll try to grab more logs, please also suggest if additional debugging flags needed. Some more observation:

* Downgrading certain packages doesn't work
  * gdm/libgdm 44.1-1 -> 44.0-1
  * gnome-shell -> 1:44.3-1 -> 1:44.2-1
* Replacing `gdm` with `xdm` or `sddm` seems to be a workaround (still running `gnome-shell`)

**Comment by shymega** [comment]
Yeah, I just checked the `PKGBUILD`.

Intriguing that there seems to be a link with `gdm`. Can you run Input Leap with the following env var set: `$MUTTER_SYNC=1`?

**Comment by shymega** [comment]
Interestingly enough, I found another issue detailing a very similar error: https://github.com/home-sweet-gnome/dash-to-panel/issues/1912

**Comment by fantix** [comment]
```
Aug 07 11:19:06 host-pc gnome-software[10615]: gs_plugin_cache_lookup: assertion 'key != NULL' failed
Aug 07 11:19:06 host-pc gnome-software[10615]: gs_plugin_cache_add: assertion 'key != NULL' failed
Aug 07 11:19:07 host-pc gnome-software[10615]: Only 10 apps for curated list, hiding
Aug 07 11:19:10 host-pc input-leap[10560]: ("-f", "--no-tray", "--debug", "INFO", "--name", "fantix-pc", "[192.168.2.222]:24800")
Aug 07 11:19:18 host-pc gnome-shell[10397]: Can't update stage views actor <unnamed>[<ShellTrayIcon>:0x55974f8f9390] is on because it needs an allocation.
Aug 07 11:19:18 host-pc gnome-shell[10397]: Can't update stage views actor <unnamed>[<ShellTrayIcon>:0x55974f8f7280] is on because it needs an allocation.
Aug 07 11:19:18 host-pc gnome-shell[10397]: Can't update stage views actor <unnamed>[<ShellTrayIcon>:0x55974f80dee0] is on because it needs an allocation.
Aug 07 11:19:18 host-pc gnome-shell[10397]: Can't update stage views actor <unnamed>[<MetaWindowActorX11>:0x55974f8fc2c0] is on because it needs an allocation.
Aug 07 11:19:18 host-pc gnome-shell[10397]: Can't update stage views actor <unnamed>[<MetaSurfaceActorX11>:0x55974f8fca50] is on because it needs an allocation.
Aug 07 11:19:18 host-pc gnome-shell[10397]: Can't update stage views actor <unnamed>[<MetaWindowActorX11>:0x55974f900020] is on because it needs an allocation.
Aug 07 11:19:18 host-pc gnome-shell[10397]: Can't update stage views actor <unnamed>[<MetaSurfaceActorX11>:0x55974f900980] is on because it needs an allocation.
Aug 07 11:19:18 host-pc gnome-shell[10397]: Can't update stage views actor <unnamed>[<MetaWindowActorX11>:0x55974f904630] is on because it needs an allocation.
Aug 07 11:19:18 host-pc gnome-shell[10397]: Can't update stage views actor <unnamed>[<MetaSurfaceActorX11>:0x55974f904d60] is on because it needs an allocation.
Aug 07 11:19:22 host-pc com.google.Chrome.desktop[11355]: [2:2:0807/111922.740883:ERROR:nacl_browser.cc(288)] Failed to open NaCl IRT file "/app/extra/nacl_irt_x86_64.nexe": -4
Aug 07 11:19:23 host-pc com.google.Chrome.desktop[11355]: [2:2:0807/111923.892638:ERROR:object_proxy.cc(590)] Failed to call method: org.freedesktop.ScreenSaver.GetActive: object_path= /org/freedesktop/ScreenSaver: org.freedesktop.DBus.Error.NotSupported: This method is not implemented
Aug 07 11:19:24 host-pc gnome-shell[10397]: Error connecting to the screencast service
```

The same error here as in the first post, then I also found:

```
Aug 07 11:19:28 host-pc kernel: gnome-shell[11725]: segfault at 28 ip 00007fec93aef3a4 sp 00007ffc4a344e70 error 4 in libmutter-clutter-12.so.0.0.0[7fec93a64000+bb000] likely on CPU 27 (core 11, socket 0)
Aug 07 11:19:28 host-pc kernel: Code: 00 48 8d 35 9e 5d 04 00 48 8d 3d 8c fc 02 00 ff 15 b1 8b 08 00 90 f3 0f 1e fa 41 55 41 54 55 48 89 f5 53 48 89 fb 48 83 ec 38 <4c> 8b 67 28 64 48 8b 04 25 28 00 00 00 48 89 44 24 28 31 c0 48 8b
```

**Comment by fantix** [comment]
> Yeah, I just checked the `PKGBUILD`.
> 
> Intriguing that there seems to be a link with `gdm`. Can you run Input Leap with the following env var set: `$MUTTER_SYNC=1`?

Let me try this

**Comment by shymega** [comment]
I think this is a low-level GNOME `mutter` issue. I found a patch on their GitLab, which merged three months ago.

I don't think this is Input Leap-specific, but it would be useful for you to build from the v2.4.0 tag, install, and test again. Make sure to uninstall the Git version from both machines - the v2.4.0 should be on both servers/clients.

This way, we can isolate the cause. The logs are indicating a missing XDG portal, and those portals are not linked to Input Leap.

**Comment by shymega** [comment]
Do you use `caffeine-ng`, or anything that interacts/enables/disables DPMS?

The patch I found is no longer relevant, but I might have found the cause for you.

**Comment by fantix** [comment]
> Do you use `caffeine-ng`

Negative.

> or anything that interacts/enables/disables DPMS?

I don't think I have, but I did turn off screensavier / screen auto-off.

```
$ xset q
...
DPMS (Display Power Management Signaling):
  Standby: 600    Suspend: 600    Off: 600
  DPMS is Disabled
```

Thank you so much for looking into my issue!

**Comment by shymega** [comment]
Hm, so it's probably not that then. I would submit a bug to Arch's bug tracker, with the logs you sent here. They may ask you to bring it further upstream to `libmutter`, but I've seen a couple of other bugs with similar errors in 2023 - recently.

**Comment by fantix** [comment]
No problem, thank you for your time!

**Comment by shymega** [comment]
@fantix No worries. It would be helpful to track the bug reports here in this issue too. So links, would be useful.

**Comment by prahal** [comment]
@fantix Could you include the link to the archlinux bug?

@shymega why set MUTTER_SYNC on input-leap (sorry I don't know much about MUTTER_SYNC).

In the gnome debug instruction I saw about MUTTER_SYNC it had to be set for gnome-shell.
Which on a systemd system (https://wiki.archlinux.org/title/Systemd/User#Environment_variables) can be achieved by creating:
```
$HOME/.config/environment.d/10-mutter-sync.conf
```
with:
```
MUTTER_SYNC=1
```
in it.

Likely input-leap might be able to avoid triggering the crash, but gnome-shell/Xorg should not crash in the first place.

**Comment by shymega** [comment]
@prahal 

> @shymega why set MUTTER_SYNC on input-leap (sorry I don't know much about MUTTER_SYNC).

I don't remember the exact thread on GitLab, but a GNOME developer suggested setting that env in order to debug the error message which was similar to the one here.

**Comment by fantix** [comment]
> @fantix Could you include the link to the archlinux bug?

Yeah sorry I unblocked myself by switching to KDE/Plasma, so I haven't got to create that upstream bug.

**Comment by prahal** [comment]
@shymega Thanks.
If told by a GNOME dev, from https://developer-old.gnome.org/meta/stable/running-mutter.html I believe this variable is meant to be added to the mutter/gnome-shell environment (only) per it affects mutter only.
But indeed it will make Xorg error synced to the gnome-shell/mutter calls.

**Comment by shymega** [comment]
> @shymega Thanks. If told by a GNOME dev, from https://developer-old.gnome.org/meta/stable/running-mutter.html I believe this variable is meant to be added to the mutter/gnome-shell environment (only) per it affects mutter only. But indeed it will make Xorg error synced to the gnome-shell/mutter calls.

Indeed. I wasn't told this, I just read about it. It's probably added confusion, me suggesting it.

**Comment by prahal** [comment]
@fantix you might want to upgrade gtk4 to 4.10.4  https://forum.manjaro.org/t/stable-update-2023-06-04-kernels-gnome-44-1-plasma-5-27-5-python-3-11-toolchain-firefox/141610/31 ( itself from https://forum.manjaro.org/t/stable-update-2023-06-04-kernels-gnome-44-1-plasma-5-27-5-python-3-11-toolchain-firefox/141610/44?page=3 )

**Comment by prahal** [comment]
@fantix could you test if you can reproduce your old issue about gnome-shelll crashing when an input leap client in Xorg mode with gnome-shell/mutter 44.5 or newer? (if only as a one-shot test)
So we can close this bug which I believe is a gnome shell issue (which could turns out to trigger only when using input leap on your setup).

**Comment by prahal** [comment]
@fantix can you still reproduce the gnome shell Xorg crash (or Wayland if you switched ?) with input leap when playing YouTube ? Per the link I provided in this thread more than a year ago, this bug should be fixed already (in mutter 44.5).

**Comment by prahal** [comment]
NEEDINFO, no reply for more than a year. Likely already fixed long ago in mutter.

If reproducible, what might help is pasting the back trace with MUTTER_SYNC=1 (it won't prevent the crash in any way but will provide a meaningful stack trace).
